napari/components/viewer_model.py

```diff
@@ -433,14 +433,17 @@ def _update_cursor_size(self, event):
 
     def _on_cursor_position_change(self, event):
         """Set the layer cursor position."""
-        for layer in self.layers:
-            layer.position = self.cursor.position
-
-    def _on_grid_change(self, event):
-        """Arrange the current layers is a 2D grid."""
         for i, layer in enumerate(self.layers[::-1]):
-            i_row, i_column = self.grid.position(i, len(self.layers))
-            self._subplot(layer, (i_row, i_column))
+            # Offset the position of the cursor if in grid mode
+            if self.grid.enabled:
+                grid_position = self.grid.position(i, len(self.layers))
+                offset = np.multiply(grid_position, self.grid.size)
+                position = list(self.cursor.position)
+                position[self.dims.displayed[-1]] -= offset[-1]
+                position[self.dims.displayed[-2]] -= offset[-2]
+                layer.position = position
```

Comment:
I kind of think that this logic should use a transform (for which we have models) rather than this custom translation. Would that be easy enough to add?